### PR TITLE
Update covidhub import to handle internal CZBID

### PR DESCRIPTION
**Pull request by ttung**
### Description
This adds support for Internal CZBIDs.

The old workflow was:

1. look at all consensus genomes.
2. look at all CZBIDs that are not represented by the consensus genomes.
3. write out entries for the consensus genomes.
4. write out entries for the failed genomes.

This refactors the workflow to:

1. Pull all the CZBIDs associtaed with the project.
2. Load all the consensus genomes.
3. Loop through all the CZBIDs and:
   a. Find the consensus genome
   b. Sort out the metadata, depending on the CZBID type.

Depends on #297 

#### Issue
[ch133318](https://app.clubhouse.io/genepi/story/133318/port-over-internal-samples-from-covidhub-db-into-appropriate-projects-about-500)

### Test plan
import marin's internal czbids (RR089i)

### CODE REVIEW

File: src/backend/aspen/covidhub_import/implementation.py [hunk 1]
```diff
@@ -120,68 +152,196 @@ def import_project(
...
+                    external_accession = f"{czbid.external_accession} ({czbid.czb_id})"
+                else:
+                    external_accession = czbid.external_accession
+                collection_date = czbid.collection_date
+                date_received = czbid.date_received or czbid.collection_date
+            elif isinstance(czbid, InternalCZBID):
+                internal_metadata = internal_czb_ids_metadata[czbid.czb_id]
+                external_accession = internal_metadata[0]
+                collection_date = internal_metadata[1]
+                date_received = internal_metadata[1]
```
Comment (phoenixAja) [inline src/backend/aspen/covidhub_import/implementation.py:273]: should this be this? 
```suggestion
                date_received = internal_metadata[2]
```

File: src/backend/aspen/covidhub_import/implementation.py [hunk 2]
```diff
@@ -103,8 +120,23 @@ def import_project(
         group: Group = session.query(Group).filter(Group.id == aspen_group_id).one()
 
         logger.info("Retrieving from COVIDHUB DB...")
-        consensus_genomes: Collection[ConsensusGenome] = (
-            covidhub_session.query(ConsensusGenome)
+        group_czbids: Collection[CZBID] = (
+            covidhub_session.query(CZBID)
+            .join(Project)
+            .filter(Project.rr_project_id == rr_project_id)
```
Comment (shanaxel42) [inline src/backend/aspen/covidhub_import/implementation.py:138]: So the rr_project id is different for internal/external samples though, ex. `RR0092i` vs `RR0092e` am I reading this wrong or does it assume they are the same?

File: src/backend/aspen/covidhub_import/implementation.py [hunk 3]
```diff
@@ -103,8 +120,23 @@ def import_project(
...
+            .options(
+                selectin_polymorphic(CZBID, [DphCZBID, InternalCZBID]),
+                joinedload(CZBID.genome_submission_info),
+            )
+            .all()
+        )
+
+        czbid_lookup_by_str: Mapping[str, CZBID] = {
+            czbid.czb_id: czbid for czbid in group_czbids
```
Comment (shanaxel42) [inline src/backend/aspen/covidhub_import/implementation.py:148]: Why do we do this str comparison?

File: src/backend/aspen/covidhub_import/implementation.py (hunk 1 above)
Comment (ttung) [reply src/backend/aspen/covidhub_import/implementation.py:273]: no, this is intentionally the same.  we don't really know the collection date, so we're setting it to `date_received`.

File: src/backend/aspen/covidhub_import/implementation.py [hunk 4]
```diff
@@ -120,68 +152,196 @@ def import_project(
...
             .join(Project)
-            .filter(
-                DphCZBID.czb_id.notin_(
-                    {
-                        consensus_genome.sample_fastqs.czb_id.czb_id
-                        for consensus_genome in consensus_genomes
-                    }
-                )
+            .join(CZBIDRnaPlate)
+            .join(
```
Comment (shanaxel42) [inline src/backend/aspen/covidhub_import/implementation.py:186]: clleeeeeean

File: src/backend/aspen/covidhub_import/implementation.py [hunk 5]
```diff
@@ -120,68 +152,196 @@ def import_project(
...
+        # verify that we have no duplicates for external accessions
+        external_accession_to_czbids: DefaultDict[
+            str, set[str]
+        ] = collections.defaultdict(set)
+        for czbid, (external_accession, _) in internal_czb_ids_metadata.items():
+            external_accession_to_czbids[external_accession].add(czbid)
+        czbids_to_process: MutableSequence[CZBID] = list()
+        for external_accession, czbids in external_accession_to_czbids.items():
+            if len(czbids) > 1:
+                # gather up the information for each czbid representing the external
```
Comment (shanaxel42) [inline src/backend/aspen/covidhub_import/implementation.py:218]: So I'm not sure we want this logic. If a county accidentally sent two samples with the same external_accession they may actually be two separate samples. Ideally there wouldn't be any of these, but are there?

File: src/backend/aspen/covidhub_import/implementation.py [hunk 6]
```diff
@@ -120,68 +152,196 @@ def import_project(
...
-            if not isinstance(czbid, DphCZBID):
-                warnings.warn(
-                    "consensus genome has non-dphczbid"
-                    f"{consensus_genome.sample_fastqs.czb_id}"
-                )
+        for czbid in tqdm.tqdm(czbids_to_process):
+            if isinstance(czbid, DphCZBID):
+                if czbid.external_accession in (
+                    "OCPHL2207",
+                    "OCPHL2208",
```
Comment (shanaxel42) [inline src/backend/aspen/covidhub_import/implementation.py]: lolz

**Review by shanaxel42 — CHANGES_REQUESTED** [review]

File: src/backend/aspen/covidhub_import/implementation.py (hunk 2 above)
Comment (ttung) [reply src/backend/aspen/covidhub_import/implementation.py:138]: We'd run the import twice -- once for `RR0092e` and once for `RR0092i`

File: src/backend/aspen/covidhub_import/implementation.py (hunk 3 above)
Comment (ttung) [reply src/backend/aspen/covidhub_import/implementation.py:148]: There is no string comparison...?
This is to look up a czbid based on the string.

File: src/backend/aspen/covidhub_import/implementation.py (hunk 4 above)
Comment (ttung) [reply src/backend/aspen/covidhub_import/implementation.py:186]: this should just be one join.

File: src/backend/aspen/covidhub_import/implementation.py (hunk 5 above)
Comment (ttung) [reply src/backend/aspen/covidhub_import/implementation.py:218]: Yes, there are.

These accessions are assigned when czb got the tubes (`AccessionSamples`), and not assigned by the counties afaik.

Comment (ttung) [reply src/backend/aspen/covidhub_import/implementation.py:218]: Example for Marin:

```
External accession H12313 represented by multiple czbids: {'RR089i_00186', 'RR089i_00181'}.  Choosing RR089i_00181 as the czbid forthis sample.
External accession H11851 represented by multiple czbids: {'RR089i_00182', 'RR089i_00187'}.  Choosing RR089i_00182 as the czbid forthis sample.
```

File: src/backend/aspen/covidhub_import/implementation.py (hunk 4 above)
Comment (ttung) [reply src/backend/aspen/covidhub_import/implementation.py:186]: fixed!

File: src/backend/aspen/covidhub_import/implementation.py (hunk 2 above)
Comment (shanaxel42) [reply src/backend/aspen/covidhub_import/implementation.py:138]: ahhhhh then all good!

File: src/backend/aspen/covidhub_import/implementation.py (hunk 5 above)
Comment (shanaxel42) [reply src/backend/aspen/covidhub_import/implementation.py:218]: Oh interesting these are internal samples that were picked twice, ok yea then this makes sense

**Review by shanaxel42 — APPROVED** [review]
Looks good!

File: src/backend/aspen/covidhub_import/implementation.py [hunk 7]
```diff
@@ -120,68 +152,196 @@ def import_project(
...
-                        for consensus_genome in consensus_genomes
-                    }
-                )
+        }
+
+        internal_czb_ids_metadata: Mapping[str, Tuple[str, datetime.datetime]] = {
+            czbid: (external_accession, collection_date)
+            for czbid, external_accession, collection_date in covidhub_session.query(
+                InternalCZBID.czb_id.label("czb_id"),
+                AccessionSample.accession.label("external_accession"),
```
Comment (shanaxel42) [inline src/backend/aspen/covidhub_import/implementation.py]: OK! so yes we want this to be

`AccessionSample.location_submitter_id.label("external_accession"),`

**Review by shanaxel42 — CHANGES_REQUESTED** [review]

File: src/backend/aspen/covidhub_import/implementation.py [hunk 8]
```diff
@@ -120,68 +166,195 @@ def import_project(
...
+        }
+
+        internal_czb_ids_metadata: Mapping[str, Tuple[str, datetime.datetime]] = {
+            czbid: (
+                _get_external_accession(external_accession, czbid_lookup_by_str[czbid]),
+                collection_date,
+            )
+            for czbid, external_accession, collection_date in covidhub_session.query(
+                InternalCZBID.czb_id.label("czb_id"),
+                AccessionSample.location_submitter_id.label("external_accession"),
```
Comment (shanaxel42) [inline src/backend/aspen/covidhub_import/implementation.py:183]: nice!

**Review by shanaxel42 — APPROVED** [review]
Looks good!

**Pull request merged**